File: DirectShowFilters/TsWriter/source/SvctParser.cpp
/* 
 *	Copyright (C) 2006-2008 Team MediaPortal
 *	http://www.team-mediaportal.com
 *
 *  This Program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 *   
 *  This Program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU General Public License for more details.
 *   
 *  You should have received a copy of the GNU General Public License
 *  along with GNU Make; see the file COPYING.  If not, write to
 *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA. 
 *  http://www.gnu.org/copyleft/gpl.html
 *
 */
#include "SvctParser.h"

extern void LogDebug(const char *fmt, ...);
extern bool DisableCRCCheck();

CSvctParser::CSvctParser(void)
{
  //SetPid(0);
  if (DisableCRCCheck())
  {
    EnableCrcCheck(false);
  }
  Reset();
  m_pCallBack = NULL;
}

CSvctParser::~CSvctParser(void)
{
}

void CSvctParser::Reset()
{
  LogDebug("SvctParser: reset");
  CSectionDecoder::Reset();
  m_mDefinedChannelMap.clear();
  m_mInverseChannelMap.clear();
  LogDebug("SvctParser: reset done");
}

void CSvctParser::SetCallBack(ISvctCallBack* callBack)
{
  m_pCallBack = callBack;
}

void CSvctParser::OnNewSection(CSection& sections)
{
  if (sections.table_id != 0xc4)
  {
    LogDebug("SvctParser: other table ID 0x%x", sections.table_id);
    return;
  }
  if (m_pCallBack == NULL)
  {
    return;
  }
  byte* section = sections.Data;

  try
  {
    int section_length = ((section[1] & 0x0f) << 8) + section[2];
    if (section_length > 1021 || section_length < 8)
    {
      LogDebug("SvctParser: invalid section length = %d", section_length);
      return;
    }
    int protocol_version = (section[3] & 0x1f);
    int transmission_medium = (section[4] >> 4);
    int table_subtype = (section[4] & 0xf);
    int vct_id = (section[5] << 8) + section[6];

    LogDebug("SvctParser: section length = %d, protocol version = %d, transmission medium = %d, table subtype = %d", section_length, protocol_version, transmission_medium, table_subtype);

    int pointer = 7;
    int subtableLength = 0;
    int endOfSection = section_length - 1;
    if (table_subtype == 0)
    {
      DecodeVirtualChannelMap(&section[7], section_length - 8, &subtableLength);
    }
    else if (table_subtype == 1)
    {
      DecodeDefinedChannelMap(&section[7], section_length - 8, &m_mDefinedChannelMap, &subtableLength);
    }
    else if (table_subtype == 2)
    {
      DecodeInverseChannelMap(&section[7], section_length - 8, &m_mInverseChannelMap, &subtableLength);
    }
    else
    {
      LogDebug("SvctParser: unsupported table subtype %d", table_subtype);
      return;   // attempting to parse descriptors after skipping the table would be bad!
    }

    pointer += subtableLength;

    while (pointer + 1 < endOfSection)
    {
      int tag = section[pointer++];
      int length = section[pointer++];
      LogDebug("SvctParser: descriptor, tag = 0x%x, length = %d, pointer = %d, end of descriptor = %d, end of section = %d, section length = %d", tag, length, pointer, pointer + length, endOfSection, section_length);
      if (pointer + length > endOfSection)
      {
        LogDebug("SvctParser: invalid descriptor length = %d, pointer = %d, end of descriptor = %d, end of section = %d, section length = %d", length, pointer, pointer + length, endOfSection, section_length);
        return;
      }

      pointer += length;
    }

    if (pointer != endOfSection)
    {
      LogDebug("SvctParser: section parsing error");
    }
  }
  catch (...)
  {
    LogDebug("SvctParser: unhandled exception in OnNewSection()");
  }
}

void CSvctParser::DecodeVirtualChannelMap(byte* b, int sectionLength, int* mapLength)
{
  const int VIRTUAL_CHANNEL_SIZE = 9;
  *mapLength = 0;
  if (sectionLength < 7)
  {
    LogDebug("SvctParser: invalid section length for virtual channel map = %d", sectionLength);
    return;
  }

  bool descriptors_included = ((b[0] & 0x20) != 0);
  bool splice = ((b[1] & 0x80) != 0);
  unsigned int activation_time = (b[2] << 24) + (b[3] << 16) + (b[4] << 8) + b[5];
  int number_of_vc_records = b[6];
  int pointer = 7;
  if ((pointer + (number_of_vc_records * VIRTUAL_CHANNEL_SIZE)) > sectionLength) // does not include descriptors - impossible to include them until we encounter them
  {
    LogDebug("SvctParser: invalid virtual channel map record count = %d, section length = %d", number_of_vc_records, sectionLength);
    return;
  }

  for (int i = 0; i < number_of_vc_records; i++)
  {
    CChannelInfo* info = NULL;
    DecodeVirtualChannel(&b[pointer], &info);
    pointer += VIRTUAL_CHANNEL_SIZE;

    if (descriptors_included)
    {
      int descriptors_count = b[pointer++];
      for (int d = 0; d < descriptors_count; d++)
      {
        if (pointer + 1 > sectionLength)
        {
          LogDebug("SvctParser: detected virtual channel descriptor count %d for channel %d is invalid in loop %d, pointer = %d, section length = %d", descriptors_count, i, d, pointer, sectionLength);
          return;
        }

        int tag = b[pointer++];
        int length = b[pointer++];
        LogDebug("SvctParser: virtual channel descriptor, tag = 0x%x, length = %d, pointer = %d, end of descriptor = %d, section length = %d", tag, length, pointer, pointer + length, sectionLength);
        if (pointer + length > sectionLength)
        {
          LogDebug("SvctParser: invalid virtual channel descriptor length = %d, pointer = %d, end of descriptor = %d, section length = %d", length, pointer, pointer + length, sectionLength);
          return;
        }

        pointer += length;
      }
    }

    if (info != NULL)
    {
      if (m_pCallBack != NULL)
      {
        m_pCallBack->OnSvctReceived(*info);
      }
      delete info;
    }
  }

  *mapLength = pointer;
}

void CSvctParser::DecodeDefinedChannelMap(byte* b, int sectionLength, map<int, bool>* dcm, int* mapLength)
{
  *mapLength = 0;
  if (sectionLength < 3)
  {
    LogDebug("SvctParser: invalid section length for defined channel map = %d", sectionLength);
    return;
  }
  if (dcm == NULL)
  {
    LogDebug("SvctParser: defined channel map pointer is NULL");
    return;
  }
  int first_virtual_channel = ((b[0] & 0xf) << 8) + b[1];
  int dcm_data_length = (b[2] & 0x7f);
  *mapLength = 3 + dcm_data_length;
  if (*mapLength > sectionLength)
  {
    LogDebug("SvctParser: invalid defined channel data length = %d, section length = %d", dcm_data_length, sectionLength);
    *mapLength = 0;
    return;
  }
  int pointer = 3;
  int currentChannel = first_virtual_channel;
  for (int i = 0; i < dcm_data_length; i++)
  {
    bool range_defined = ((b[pointer] & 0x80) != 0);
    int channels_count = (b[pointer++] & 0x7f);

    for (int c = 0; c < channels_count; c++)
    {
      (*dcm)[currentChannel++] = range_defined;
    }
  }
}

void CSvctParser::DecodeInverseChannelMap(byte* b, int sectionLength, map<int, int>* icm, int* mapLength)
{
  *mapLength = 0;
  if (sectionLength < 3)
  {
    LogDebug("SvctParser: invalid section length for inverse channel map = %d", sectionLength);
    return;
  }
  if (icm == NULL)
  {
    LogDebug("SvctParser: inverse channel map pointer is NULL");
    return;
  }
  int first_map_index = ((b[0] & 0xf) << 8) + b[1];
  int record_count = (b[2] & 0x7f);
  *mapLength = 3 + (record_count * 4);
  if (*mapLength > sectionLength)
  {
    LogDebug("SvctParser: invalid inverse channel record count = %d, section length = %d", record_count, sectionLength);
    *mapLength = 0;
    return;
  }
  int pointer = 3;
  for (int i = 0; i < record_count; i++)
  {
    int source_id = (b[pointer] << 8) + b[pointer + 1];
    pointer += 2;
    int virtual_channel_number = ((b[pointer] & 0x0f) << 8) + b[pointer + 1];
    pointer += 2;

    (*icm)[source_id] = virtual_channel_number;
  }
}

void CSvctParser::DecodeVirtualChannel(byte* b, CChannelInfo** info)
{
  int pointer = 0;
  int virtual_channel_number = ((b[pointer] & 0x0f) << 8) + b[pointer + 1];
  pointer += 2;
  bool application_virtual_channel = ((b[pointer] & 0x80) != 0);
  int path_select = ((b[pointer] >> 5) & 1);
  int transport_type = ((b[pointer] >> 4) & 1);
  int channel_type = (b[pointer++] & 0xf);
  int source_id = (b[pointer] << 8) + b[pointer + 1];
  pointer += 2;
  LogDebug("SvctParser: virtual channel, number = %d, is application = %d, path select = %d, transport type = %d, channel type = %d, source ID = 0x%x",
            virtual_channel_number, application_virtual_channel, path_select, transport_type, channel_type, source_id);

  int cds_reference = b[pointer++];
  int program_number = 0;
  int mms_reference = 0;
  bool scrambled = false;
  int video_standard = 0;
  if (transport_type == 0)  // MPEG-2 transport
  {
    program_number, mms_reference, 
    program_number = (b[pointer] << 8) + b[pointer + 1];
    pointer += 2;
    mms_reference = b[pointer++];
    LogDebug("SvctParser:   CDS reference = 0x%x, program number = 0x%x, MMS reference = 0x%x", cds_reference, program_number, mms_reference);
  }
  else  // (analog)
  {
    scrambled = ((b[pointer] & 0x80) != 0);
    video_standard = (b[pointer++] & 0x0f);
    pointer += 2;
    LogDebug("SvctParser:   CDS reference = 0x%x, scrambled = %d, video standard = %d", cds_reference, scrambled, video_standard);
  }

  // If the channel is analog, hidden, on an alternative path, or is an application, then the channel
  // is inaccessible or not meant to be shown to the user.
  if (transport_type != 0 || channel_type != 0 || path_select != 0 || application_virtual_channel)
  {
    LogDebug("SvctParser: not a user channel");
    return;
  }

  *info = new CChannelInfo();
  (*info)->LCN = virtual_channel_number;
  (*info)->NetworkId = source_id;
  (*info)->ServiceId = program_number;
}

File: TvEngine3/TVLibrary/TVLibrary/Implementations/Dri/Parser/SvctParser.cs
#region Copyright (C) 2005-2011 Team MediaPortal

// Copyright (C) 2005-2011 Team MediaPortal
// http://www.team-mediaportal.com
// 
// MediaPortal is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 2 of the License, or
// (at your option) any later version.
// 
// MediaPortal is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with MediaPortal. If not, see <http://www.gnu.org/licenses/>.

#endregion

using System;

namespace TvLibrary.Implementations.Dri.Parser
{
  public class SvctParser
  {
    private enum TableSubtype
    {
      VirtualChannelMap = 0,
      DefinedChannelMap,
      InverseChannelMap
    }

    private enum TransportType
    {
      Mpeg2,
      NonMpeg2
    }

    private enum ChannelType
    {
      Normal,
      Hidden
    }

    private enum VideoStandard
    {
      Ntsc = 0,
      Pal625,
      Pal525,
      Secam,
      Mac
    }

    public void Decode(byte[] section)
    {
      if (section == null || section.Length < 13)
      {
        return;
      }

      byte tableId = section[2];
      if (tableId != 0xc4)
      {
        return;
      }
      int sectionLength = ((section[3] & 0x0f) << 8) + section[4];
      if (section.Length != 2 + sectionLength + 3)
      {
        Log.Log.Error("S-VCT: invalid section length = {0}, byte count = {1}", sectionLength, section.Length);
        return;
      }
      byte protocolVersion = (byte)(section[5] & 0x1f);
      byte transmissionMedium = (byte)(section[6] >> 4);
      TableSubtype tableSubtype = (TableSubtype)(section[6] & 0x0f);
      int vctId = (section[7] << 8) + section[8];
      Log.Log.Debug("S-VCT: protocol version = {0}, transmission medium = {1}, table subtype = {2}, VCT ID = 0x{3}",
        protocolVersion, transmissionMedium, tableSubtype, vctId);

      int pointer = 9;
      int endOfSection = section.Length - 4;
      try
      {
        switch (tableSubtype)
        {
          case TableSubtype.DefinedChannelMap:
            DecodeDefinedChannelMap(section, endOfSection, ref pointer);
            break;
          case TableSubtype.VirtualChannelMap:
            DecodeVirtualChannelMap(section, endOfSection, ref pointer);
            break;
          case TableSubtype.InverseChannelMap:
            DecodeInverseChannelMap(section, endOfSection, ref pointer);
            break;
          default:
            Log.Log.Error("S-VCT: unsupported subtable type {0}", tableSubtype);
            return;
        }
      }
      catch (Exception ex)
      {
        Log.Log.Error(ex.Message);
        return;
      }

      while (pointer + 1 < endOfSection)
      {
        byte tag = section[pointer++];
        byte length = section[pointer++];
        Log.Log.Debug("S-VCT: descriptor, tag = 0x{0:x}, length = {1}", tag, length);
        if (pointer + length > endOfSection)
        {
          Log.Log.Error("S-VCT: invalid descriptor length, pointer = {0}, end of section = {1}, descriptor length = {2}", pointer, endOfSection, length);
          return;
        }

        if (tag == 0x93)  // revision detection descriptor
        {
          ParserCommon.DecodeRevisionDetectionDescriptor(section, pointer, length);
        }

        pointer += length;
      }

      if (pointer != endOfSection)
      {
        Log.Log.Error("S-VCT: corruption detected at end of section, pointer = {0}, end of section = {1}", pointer, endOfSection);
        return;
      }
    }

    private void DecodeDefinedChannelMap(byte[] section, int endOfSection, ref int pointer)
    {
      if (pointer + 3 > endOfSection)
      {
        throw new Exception(string.Format("S-VCT: corruption detected at defined channel map, pointer = {0}, end of section = {1}", pointer, endOfSection));
      }

      int firstVirtualChannel = ((section[pointer] & 0x0f) << 8) + section[pointer + 1];
      pointer += 2;
      int dcmDataLength = (section[pointer++] & 0x7f);
      if (pointer + dcmDataLength > endOfSection)
      {
        throw new Exception(string.Format("S-VCT: invalid defined channel map data length {0}, pointer = {1}, end of section = {2}", dcmDataLength, pointer, endOfSection));
      }
      Log.Log.Debug("S-VCT: defined channel map, first virtual channel = {0}, DCM data length = {1}", firstVirtualChannel, dcmDataLength);
      for (int i = 0; i < dcmDataLength; i++)
      {
        bool rangeDefined = ((section[pointer] & 0x80) != 0);
        int channelsCount = (section[pointer++] & 0x7f);
        Log.Log.Debug("S-VCT: range defined = {0}, channels count = {1}", rangeDefined, channelsCount);
      }
    }

    private void DecodeVirtualChannelMap(byte[] section, int endOfSection, ref int pointer)
    {
      if (pointer + 7 > endOfSection)
      {
        throw new Exception(string.Format("S-VCT: corruption detected at virtual channel map, pointer = {0}, end of section = {1}", pointer, endOfSection));
      }

      bool descriptorsIncluded = ((section[pointer++] & 0x20) != 0);
      bool splice = ((section[pointer++] & 0x80) != 0);
      uint activationTime = 0;
      for (byte b = 0; b < 3; b++)
      {
        activationTime = activationTime << 8;
        activationTime = section[pointer++];
      }
      byte numberOfVcRecords = section[pointer++];
      Log.Log.Debug("S-VCT: virtual channel map, descriptors included = {0}, splice = {1}, activation time = {2}, number of VC records = {3}",
        descriptorsIncluded, splice, activationTime, numberOfVcRecords);

      for (byte i = 0; i < numberOfVcRecords; i++)
      {
        if (pointer + 9 > endOfSection)
        {
          throw new Exception(string.Format("S-VCT: detected number of virtual channel records {0} is invalid, pointer = {1}, end of section = {2}, loop = {3}", numberOfVcRecords, pointer, endOfSection, i));
        }

        int virtualChannelNumber = ((section[pointer] & 0x0f) << 8) + section[pointer + 1];
        pointer += 2;
        bool applicationVirtualChannel = ((section[pointer] & 0x80) != 0);
        int pathSelect = ((section[pointer] & 0x20) >> 5);
        TransportType transportType = (TransportType)((section[pointer] & 0x10) >> 4);
        ChannelType channelType = (ChannelType)(section[pointer++] & 0x0f);
        int sourceId = (section[pointer] << 8) + section[pointer + 1];
        pointer += 2;
        byte cdsReference = section[pointer++];
        int programNumber = 0;
        byte mmsReference = 0;
        bool scrambled = false;
        VideoStandard videoStandard = VideoStandard.Ntsc;
        if (transportType == TransportType.Mpeg2)
        {
          programNumber = (section[pointer] << 8) + section[pointer + 1];
          pointer += 2;
          mmsReference = section[pointer++];
        }
        else
        {
          scrambled = ((section[pointer] & 0x80) != 0);
          videoStandard = (VideoStandard)(section[pointer++] & 0x0f);
          pointer += 2;
        }
        Log.Log.Debug("S-VCT: virtual channel number = {0}, application virtual channel = {1}, path select = {2}, transport type = {3}, channel type = {4}, source ID = 0x{5:x}, CDS reference = {6}, program number = 0x{7:x}, MMS reference = {8}, scrambled = {9}, video standard = {10}",
          virtualChannelNumber, applicationVirtualChannel, pathSelect, transportType, channelType, sourceId, cdsReference, programNumber, mmsReference, scrambled, videoStandard);

        if (descriptorsIncluded)
        {
          if (pointer >= endOfSection)
          {
            throw new Exception(string.Format("S-VCT: invalid section length at virtual channel map loop {0}, pointer = {0}, end of section = {1}, loop = {2}", pointer, endOfSection, i));
          }
          byte descriptorCount = section[pointer++];
          for (byte d = 0; d < descriptorCount; d++)
          {
            if (pointer + 2 > endOfSection)
            {
              throw new Exception(string.Format("S-VCT: detected virtual channel map descriptor count {0} is invalid, pointer = {1}, end of section = {2}, loop = {3}, inner loop = {4}", descriptorCount, pointer, endOfSection, i, d));
            }
            byte tag = section[pointer++];
            byte length = section[pointer++];
            Log.Log.Debug("S-VCT: virtual channel map descriptor, tag = 0x{0:x}, length = {1}", tag, length);
            if (pointer + length > endOfSection)
            {
              throw new Exception(string.Format("S-VCT: invalid virtual channel map descriptor length {0}, pointer = {1}, end of section = {2}, loop = {3}, inner loop = {4}", length, pointer, endOfSection, i, d));
            }
            pointer += length;
          }
        }
      }
    }

    private void DecodeInverseChannelMap(byte[] section, int endOfSection, ref int pointer)
    {
      if (pointer + 3 > endOfSection)
      {
        throw new Exception(string.Format("S-VCT: corruption detected at inverse channel map, pointer = {0}, end of section = {1}", pointer, endOfSection));
      }

      int firstMapIndex = ((section[pointer] & 0x0f) << 8) + section[pointer + 1];
      pointer += 2;
      int recordCount = (section[pointer++] & 0x7f);
      if (pointer + (recordCount * 4) > endOfSection)
      {
        throw new Exception(string.Format("S-VCT: invalid inverse channel map record count {0}, pointer = {1}, end of section = {2}", recordCount, pointer, endOfSection));
      }
      Log.Log.Debug("S-VCT: inverse channel map, first map index = {0}, record count = {1}", firstMapIndex, recordCount);
      for (int i = 0; i < recordCount; i++)
      {
        int sourceId = (section[pointer] << 8) + section[pointer + 1];
        pointer += 2;
        int virtualChannelNumber = ((section[pointer] & 0x0f) << 8) + section[pointer + 1];
        pointer += 2;
        Log.Log.Debug("S-VCT: source ID = 0x{0:x}, virtual channel number = {2}", sourceId, virtualChannelNumber);
      }
    }
  }
}
